edx_proctoring/api.py

```diff
@@ -494,6 +496,93 @@ def remove_allowance_for_user(exam_id, user_id, key):
         emit_event(exam, 'allowance.deleted', override_data=data)
 
 
+def add_bulk_allowances(exam_ids, user_ids, allowance_type, value):
+    """
+    Adds (or updates) an allowance for multiple users and exams
+    """
+    exam_ids = set(exam_ids)
+    user_ids = set(user_ids)
+
+    log_message = (
+        'Adding allowances of type allowance_type={allowance_type} with value={value} '
+        'for exams exam_ids={exam_ids} and users user_ids={user_ids}'.format(
+            allowance_type=allowance_type,
+            value=value,
+            exam_ids=exam_ids,
+            user_ids=user_ids
+        )
+    )
+    log.info(log_message)
+    if allowance_type == constants.TIME_MULTIPLIER:
+        err_msg = (
+            u'allowance_value "{value}" should be a float value greater than 1.'
+        ).format(value=value)
+        try:
+            multiplier = float(value) - 1
+        except ValueError:
+            raise AllowanceValueNotAllowedException(err_msg)
+        if multiplier >= 0:
+            for exam_id in exam_ids:
+                try:
+                    target_exam = get_exam_by_id(exam_id)
+                except ProctoredExamNotFoundException:
+                    log_message = (
+                        'Attempted to get exam_id={exam_id}, but this exam does not exist.'.format(exam_id=exam_id)
+                    )
+                    log.error(log_message)
+                    continue
+
+                exam_time = target_exam["time_limit_mins"]
+                added_time = round(exam_time * multiplier)
+                time_allowed = str(added_time)
+                for user_id in user_ids:
+                    try:
+                        add_allowance_for_user(exam_id, user_id,
+                                               ProctoredExamStudentAllowance.ADDITIONAL_TIME_GRANTED,
+                                               time_allowed)
+                    except ProctoredBaseException:
+                        log_message = (
+                            'Failed to add allowance for user_id={user_id} '
+                            'for exam_id={exam_id}'.format(
+                                user_id=user_id,
+                                exam_id=exam_id
+                            )
+                        )
+                        log.error(log_message)
+        else:
+            raise AllowanceValueNotAllowedException(err_msg)
+    else:
+        err_msg = (
+            u'allowance_value "{value}" should be non-negative integer value.'
+        ).format(value=value)
+        if value.isdigit():
```

Comment:
This should be checked already by the same validation logic checked in the beginning of the function.

Reply:
I was checking in both blocks of code for value validation because for the multiplier the value can be a float, however for the additional minutes the value must be an int. If I were to just check if the value was either a float or an int, then when I move into the additional time function I would have to check again to make sure that the value is an int. I was thinking I could leave it to the add_allowance_function, but the trade-off would then be that if the value is invalid for add_allowance we would still have to go through the for loop for every student and exam and receive the value error for each one rather than once in the start.

Reply:
If you read the [add_allowance_for_user](https://github.com/edx/edx-proctoring/blob/ebd87ee7759f3e8c842ef5e4706b9221efc2c577/edx_proctoring/models.py#L659) definition closely, you can see the validation of the value is already preformed there. This model function is the function called from the api function [here](https://github.com/edx/edx-proctoring/blob/84df549ba644d7a15556aa1c1281b2ab91a8a524/edx_proctoring/api.py#L444). AllowanceValueNotAllowedException is the exception raised and you did handle it. 
Notice the model field for `value` is actually a [character field](https://github.com/edx/edx-proctoring/blob/ebd87ee7759f3e8c842ef5e4706b9221efc2c577/edx_proctoring/models.py#L614), not a numbers field. 
All of this tells me your unit test cases are not covering for enough edge cases. I'll leave some comment there to prompt more test cases.